libyul/backends/evm/SSACFGLiveness.cpp

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/backends/evm/SSACFGLiveness.h>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/reverse.hpp>
+
+using namespace solidity::yul;
+
+SSACFGLiveness::SSACFGLiveness(SSACFG const& _cfg):
+	m_cfg(_cfg),
+	m_topologicalSort(_cfg),
+	m_loopNestingForest(m_topologicalSort),
+	m_liveIns(_cfg.numBlocks()),
+	m_liveOuts(_cfg.numBlocks()),
+	m_operationLiveOuts(_cfg.numBlocks())
+{
+	runDagDfs();
+	for (auto const loopRootNode: m_loopNestingForest.loopRootNodes())
+		runLoopTreeDfs(loopRootNode);
+
+	fillOperationsLiveOut();
+}
+
+void SSACFGLiveness::runDagDfs()
+{
+	// SSA Book, Algorithm 9.2
+	for (auto const blockIdValue: m_topologicalSort.postOrder())
+	{
+		// post-order traversal
+		SSACFG::BlockId blockId{blockIdValue};
+		auto const filterLiterals = [this](auto const& valueId){
+			return !std::holds_alternative<SSACFG::LiteralValue>(m_cfg.valueInfo(valueId));
+		};
+		auto const& block = m_cfg.block(blockId);
+
+		// live <- PhiUses(B)
+		std::set<SSACFG::ValueId> live{};
+		block.forEachExit(
+			[&](SSACFG::BlockId const& _successor)
+			{
+				for (auto const& phi: m_cfg.block(_successor).phis)
+				{
+					auto const& info = m_cfg.valueInfo(phi);
+					yulAssert(std::holds_alternative<SSACFG::PhiValue>(info), "value info of phi wasn't PhiValue");
+					auto const& entries = m_cfg.block(std::get<SSACFG::PhiValue>(info).block).entries;
+					// this is getting the argument index of the phi function corresponding to the path going
+					// through "blockId", ie, the currently handled block
+					auto const it = entries.find(blockId);
+					yulAssert(it != entries.end());
+					auto const argIndex = static_cast<size_t>(std::distance(entries.begin(), it));
+					if (argIndex < std::get<SSACFG::PhiValue>(info).arguments.size())
+					{
+						// todo not sure why arg index would exceed phi arg length but it does happen
+						auto const arg = std::get<SSACFG::PhiValue>(info).arguments.at(argIndex);
+						if (!std::holds_alternative<SSACFG::LiteralValue>(m_cfg.valueInfo(arg)))
+							live.insert(arg);
+					}
+					else
+						yulAssert(false);
+				}
+			});
+
+		// for each S \in succs(B) s.t. (B, S) not a back edge: live <- live \cup (LiveIn(S) - PhiDefs(S))
+		block.forEachExit(
+			[&](SSACFG::BlockId const& _successor) {
+				if (!m_topologicalSort.backEdge(blockId, _successor))
+					live += m_liveIns[_successor.value] - m_cfg.block(_successor).phis;
+			});
+		if (std::holds_alternative<SSACFG::BasicBlock::FunctionReturn>(block.exit))
+			live += std::get<SSACFG::BasicBlock::FunctionReturn>(block.exit).returnValues
+					| ranges::views::filter(filterLiterals);
+
+		// clean out unreachables
+		live = live | ranges::views::filter([&](auto const& valueId) { return !std::holds_alternative<SSACFG::UnreachableValue>(m_cfg.valueInfo(valueId)); }) | ranges::to<std::set>;
+
+		// LiveOut(B) <- live
+		m_liveOuts[blockId.value] = live;
```

Comment:
I would also suggest adding a comment to the `m_liveOuts` definition in the same way as previously suggested. For example, `m_liveOuts` contains the set of variables for each block that must be live for its successors. This includes the live-ins of a successor block (excluding those defined in a phi function of the successor) or variables used in a phi function of the successor.